Sources/NIO/SocketChannel.swift

```diff
@@ -489,36 +500,94 @@ final class DatagramChannel: BaseSocketChannel<Socket> {
+                guard self.isOpen else {
+                    throw ChannelError.eof
+                }
+                buffer.clear()
+
+                var controlByteSlice = controlBytes[...]
+                var controlMessageReceiver = ControlMessageReceiver()
+                
+                let result = try buffer.withMutableWritePointer {
+                    try self.socket.recvmsg(pointer: $0, storage: &rawAddress, storageLen: &rawAddressLength,
+                                            controlBytes: &controlByteSlice,
+                                            controlMessageReceiver: { (controlMessage) in controlMessageReceiver.receiveMessage(controlMessage) })
+                }
+                switch result {
+                case .processed(let bytesRead):
+                    assert(bytesRead > 0)
+                    assert(self.isOpen)
+                    let mayGrow = recvAllocator.record(actualReadBytes: bytesRead)
+                    readPending = false
+
+                    let msg = AddressedEnvelope(remoteAddress: rawAddress.convert(),
+                                                data: buffer,
+                                                metadata: self.reportExplicitCongestionNotifications ?
+                                                    AddressedEnvelope.Metadata(ecnState: controlMessageReceiver.ecnValue) :
+                                                    nil)
+                    assert(self.isActive)
+                    pipeline.fireChannelRead0(NIOAny(msg))
+                    if mayGrow && i < maxMessagesPerRead {
+                        buffer = recvAllocator.buffer(allocator: allocator)
+                    }
+                    readResult = .some
+                case .wouldBlock(let bytesRead):
+                    assert(bytesRead == 0)
+                    return readResult
+                }
             }
-            switch result {
-            case .processed(let bytesRead):
-                assert(bytesRead > 0)
-                assert(self.isOpen)
-                let mayGrow = recvAllocator.record(actualReadBytes: bytesRead)
-                readPending = false
-
-                let msg = AddressedEnvelope(remoteAddress: rawAddress.convert(), data: buffer)
-                assert(self.isActive)
-                pipeline.fireChannelRead0(NIOAny(msg))
-                if mayGrow && i < maxMessagesPerRead {
-                    buffer = recvAllocator.buffer(allocator: allocator)
+            return readResult
+        }
+    }
+    
+    private struct ControlMessageReceiver {
+        var ecnValue: NIOExplicitCongestionNotificationState = .transportNotCapable // Default
+        
+        mutating func receiveMessage(_ controlMessage: Socket.ControlMessage) {
+            #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+            let ipv4TosType = IP_RECVTOS
+            #else
+            let ipv4TosType = IP_TOS    // Linux
+            #endif
```

Comment:
Let's make this a `private static let` just to pull it out of the main method body.